src/hclust.jl

```diff
@@ -316,6 +316,137 @@ function hclust2{T<:Real}(d::Symmetric{T}, method::Function)
+    else
+        error("method ", method, " is unsupported by hclust_lw().")
+    end
+
+    merges = -collect(1:nc)
+    membr = ones(Float64, nc)
+    checkme = trues(nc)
+    nearestneighbor = zeros(Int64, nc-1)
+    distancetonearestneighbor = zeros(Float64, nc-1)
+
+    jm = 0; im = 0; jj = 0;
+
+    # create list of nearest neighbors
+    for i=1:(nc-1)
+        dmin = Inf
+        for j=(i+1):nc
+            ind = triindex(nc,i,j)
+            if dscpy[ind] < dmin
+                dmin = dscpy[ind]
+                jm = j
+            end
+        end
+        nearestneighbor[i] = jm
+        distancetonearestneighbor[i] = dmin
+    end
+
+    # main clustering loop
+    for nclust = nc:-1:2
+
+        # check list of nearest neighbors to determine next merge
+        dmin = Inf
+        for i=1:(nc-1)
+            if (checkme[i]==true)
+                if (distancetonearestneighbor[i] < dmin)
+                    dmin = distancetonearestneighbor[i]
+                    im = i
+                    jm = nearestneighbor[i]
+                end
+            end
+        end
+
+        # merge
+        i2 = min(im,jm) # lower element in the clustering pair
+        j2 = max(im,jm) # upper element in the clustering pair
+
+        # do merge
+        mr[nc-nclust+1] = merges[i2]
+        mc[nc-nclust+1] = merges[j2]
+        h[nc-nclust+1] = dmin
+
+        merges[i2] = nc-nclust+1
+        merges[j2] = merges[nclust]
+
+        checkme[j2] = false
+
+        # update dscpy matrix for new cluster
+        for k=1:nc
+            ind3=triindex(nc,i2,j2)
+            if (checkme[k] && (k!=i2))
+                # Lance-Williams updating
+                if (i2<k)
+                    ind1 = triindex(nc,i2,k)
+                else
+                    ind1 = triindex(nc,k,i2)
+                end
+                if (j2<k)
+                    ind2 = triindex(nc,j2,k)
+                else
+                    ind2 = triindex(nc,k,j2)
+                end
+                dscpy[ind1] = ((membr[i2] + membr[k])*dscpy[ind1] + (membr[j2] + membr[k])*dscpy[ind2] - membr[k]*dscpy[ind3]) / (membr[i2]+membr[j2]+membr[k])
+            end
+        end
+
+        # update Lance-Williams coefficients
+        membr[i2] = membr[i2] + membr[j2]
+
+        # update list of nearest neighbors
+        for i=1:(nc-1)
+            if (checkme[i]==true)
+                dmin = Inf
+                for j=(i+1):nc
+                    if (checkme[j]==true)
+                        ind = triindex(nc,i,j)
+                        if dscpy[ind] < dmin
+                            dmin = dscpy[ind]
+                            jj = j
+                        end
+                    end
+                end
+                nearestneighbor[i] = jj
+                distancetonearestneighbor[i] = dmin
+            end
+        end
+
+    end # main clustering loop
+
+    # take root if we use ward2
+    if (method == :ward2)
+        h=sqrt.(h)
```

Comment:
Use `h .= sqrt.(h)` to avoid making a copy (and possibly introducing a type instability).